python/cudf/cudf/tests/test_json.py

```diff
@@ -687,6 +687,58 @@ def test_json_types_data():
     assert df.to_arrow().equals(pa_table_pdf)
 
 
+@pytest.mark.parametrize(
+    "col_type,json_str",
+    [
+        # without quotes
+        ("int", '[{"k": 1}, {"k": 2}, {"k": 3}, {"k": 4}]'),
+        # with quotes
+        ("int", '[{"k": "1"}, {"k": "2"}]'),
+        # with quotes, mixed
+        ("int", '[{"k": "1"}, {"k": "2"}, {"k": 3}, {"k": 4}]'),
+        # with quotes, null, mixed
+        ("int", '[{"k": "1"}, {"k": "2"}, {"k": null}, {"k": 4}]'),
+        # without quotes, null
+        ("int", '[{"k": 1}, {"k": 2}, {"k": null}, {"k": 4}]'),
+        # without quotes
+        ("float", '[{"k": 1}, {"k": 2}, {"k": 3}, {"k": 4}]'),
+        # with quotes
+        ("float", '[{"k": "1"}, {"k": "2"}]'),
+        # with quotes, mixed
+        ("float", '[{"k": "1"}, {"k": "2"}, {"k": 3}, {"k": 4}]'),
+        # with quotes, null, mixed
+        ("float", '[{"k": "1"}, {"k": "2"}, {"k": null}, {"k": 4}]'),
+        # with quotes, NAN
+        ("float", '[{"k": "1"}, {"k": "2"}, {"k": NaN}, {"k": "4"}]'),
+        # without quotes
+        ("str", '[{"k": 1}, {"k": 2}, {"k": 3}, {"k": 4}]'),
+        # with quotes
+        ("str", '[{"k": "1"}, {"k": "2"}]'),
+        # with quotes, mixed
+        ("str", '[{"k": "1"}, {"k": "2"}, {"k": 3}, {"k": 4}]'),
+        # with quotes, null, mixed
+        ("str", '[{"k": "1"}, {"k": "2"}, {"k": null}, {"k": 4}]'),
+        # without quotes, null
+        ("str", '[{"k": 1}, {"k": 2}, {"k": null}, {"k": 4}]'),
+    ],
+)
+def test_json_quoted_values_with_schema(col_type, json_str):
+    experimental_df = cudf.read_json(
+        StringIO(json_str),
+        engine="cudf_experimental",
+        orient="records",
+        dtype={"k": col_type},
+    )
+    cudf_df = cudf.read_json(
+        StringIO(json_str.replace(",", "\n")[1:-1]),
+        engine="cudf",
+        orient="records",
+        lines=True,
+        dtype={"k": col_type},
+    )
+    assert_eq(cudf_df, experimental_df)
```

Comment:
Great suggestion. Compared it. 
pandas has different behaviour than both cudf engine.

I think pandas infers types and parses the json, and then converts to the user specified type. Because of this, there are some difference in output.
1) All int literals:
For example, for json string `[{"k": 1}, {"k": 2}, {"k": null}, {"k": 4}]` 
`pd.read_json(StringIO(jsonstr), orient="records", dtype={"k": "str"})` creates output
```
     k
0  1.0
1  2.0
2  NaN
3  4.0
```
while both cudf engines create output
```
      k
0     1
1     2
2  <NA>
3     4
```

2) int and `"` string literals mixed
If quotes are present, even if there are literals and user specified type is `int`, it parses as string. This seems  very odd behaviour compared to previous example.

Reply:
Should we deviate from pandas behaviour, or replicate the pandas behaviour in cudf?

Reply:
Oof. To put it politely, our behavior makes more sense to me. Let's keep it at least until a customer asks for different behavior :D 
I still wish the test would compare with ground truth. Maybe add a test with one or few cases where the output is compared with hard-coded values?

Reply:
added ground truth test cases.